pkg/operation/shoot/shoot.go

```diff
@@ -466,6 +469,31 @@ func (s *Shoot) IsShootControlPlaneLoggingEnabled(c *config.GardenletConfigurati
 	return s.Purpose != gardencorev1beta1.ShootPurposeTesting && gardenlethelper.IsLoggingEnabled(c)
 }
 
+// IsHAControlPlaneConfigured returns true if HA configuration for the shoot control plane has been set either
+// via an alpha-annotation or ControlPlane Spec
+func (s *Shoot) IsHAControlPlaneConfigured() bool {
+	return metav1.HasAnnotation(s.GetInfo().ObjectMeta, v1beta1constants.ShootAlphaControlPlaneHighAvailability) || s.GetInfo().Spec.ControlPlane != nil
+}
+
+// GetFailureToleranceType determines the FailureToleranceType by looking at both the alpha HA annotations and shoot spec ControlPlane
+func (s *Shoot) GetFailureToleranceType() *gardencorev1beta1.FailureToleranceType {
+	if gardenletfeatures.FeatureGate.Enabled(features.HAControlPlanes) {
+		if haAnnot, ok := s.GetInfo().ObjectMeta.Annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]; ok {
+			var failureToleranceType gardencorev1beta1.FailureToleranceType
+			if haAnnot == v1beta1constants.ShootAlphaControlPlaneHighAvailabilityMultiZone {
+				failureToleranceType = gardencorev1beta1.FailureToleranceTypeZone
+			} else {
+				failureToleranceType = gardencorev1beta1.FailureToleranceTypeNode
+			}
+			return &failureToleranceType
+		}
+		if s.GetInfo().Spec.ControlPlane != nil {
+			return &s.GetInfo().Spec.ControlPlane.HighAvailability.FailureTolerance.FailureToleranceType
+		}
+	}
```

Comment:
Move this to the API helper package? Then you can re-use it in the scheduler instead of duplicating the code?

Reply:
I cannot move this as this operates on shoot struct that is defined in `types.go`. Scheduler uses `gardencorev1beta.Shoot`

Reply:
https://github.com/gardener/gardener/pull/6530#discussion_r964400079

Reply:
ok. Did not realize that shoot.GetInfo returns gardencorev1beta1.Shoot. Will move it

Reply:
If extensions might need this then we should move it already now so that we don't have revendoring delays later on. If they won't need it, it doesn't matter much.

Reply:
done

Reply:
Currently `make check` fails because of the import restrictions that are enforced via `import-boss`. Now `helper.go` has a dependency on packages which are not allowed as per `.import-restrictions` file.
```
errors in package "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper":
the following imports did not match any allowed prefix:
  github.com/gardener/gardener/pkg/features
  github.com/gardener/gardener/pkg/gardenlet/features
  k8s.io/component-base/featuregate
```
There are other packages that also use these packages. For complete list of issues see:
[check-imports-err.log](https://github.com/gardener/gardener/files/9504334/check-imports-err.log)

In the case of `helper.go` i can remove that check and move it to the caller but that is not very nice as it introduces the possibility of missed check for the feature gate and it has to be repeated everywhere. Other way is to allow this package dependency.

Reply:
Keep the feature gate handling outside in the calling packages, this should resolve it

Reply:
Since this is only used inside `botanist` it would nice if we could identify a place where helpers are kept but i am unable to locate any for now. Otherwise in all places inside botanist i will have to do the following and violate DRY:
```go
func GetFailureToleranceType(shoot *gardencorev1beta1.Shoot) *gardencorev1beta1.FailureToleranceType {
    if gardenletfeatures.FeatureGate.Enabled(features.HAControlPlanes) {
             return gardencorev1beta1helper.GetFailureToleranceType(b.Shoot.GetInfo())
    }
    return nil
}
```

Reply:
I would still keep the API-related parts of this function in the API helper package. You can still have a function on `Botanist` or `Shoot` object to call it only when the feature gate is enabled.

Reply:
done